enclave/src/lib.rs

```diff
@@ -210,27 +212,123 @@ fn update_rsa_key(path: &str) -> Result<(), sgx_status_t> {
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn init_rsa_key() -> sgx_status_t {
-    let rsa_key_path = "RSA_KEY";
+fn generate_quote() -> Option<Vec<u8>>{
+    let mut rsa_key_data: RSAKeyData = (*RSA_KEY_DATA.lock().unwrap()).clone().unwrap();
 
-    match read_rsa_key(rsa_key_path) {
+    let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs();
+    // 5 minutes before the deadline still needs to update the key
+    if (rsa_key_data.not_befer < now as i64) | (rsa_key_data.not_after - ADVANCE_REFRESH_TIME < now as i64) {
+        println!("[ENCLAVE] Rsa key not in time range. Generate new RSA key as quote customization data");
+        match update_rsa_key(RSA_KEY_PATH) {
+            Ok(_) => {
+                rsa_key_data = (*RSA_KEY_DATA.lock().unwrap()).clone().unwrap()
+            },
+            Err(_) => return None
+        };
+    }
+
+    let rsa3072_public_key = sgx_rsa3072_public_key_t {
+        modulus: rsa_key_data.modulus.clone().try_into().unwrap(),
+        exponent: rsa_key_data.e.clone().try_into().unwrap(),
+    };
+
+    let mut rt : sgx_status_t = sgx_status_t::SGX_ERROR_UNEXPECTED;
+    let mut ti : sgx_target_info_t = sgx_target_info_t::default();
+
+    let result = unsafe {
+        ocall_get_target_info(
+            &mut rt as *mut sgx_status_t,
+            &mut ti as *mut _)
+    };
+
+    if result != sgx_status_t::SGX_SUCCESS {
+        return None;
+    }
+
+    let mut public_key: Vec<u8> = Vec::new();
+    public_key.extend(&rsa3072_public_key.modulus);
+    public_key.extend(&rsa3072_public_key.exponent);
+
+    let hash_public_key = rsgx_sha256_slice(public_key.as_slice()).unwrap();
+    let mut data = [0; 64];
+    data[..hash_public_key.len()].copy_from_slice(&hash_public_key);
+    println!("[+] get Hash(rsa public key) for report");
+    let report_data: sgx_report_data_t = sgx_report_data_t {
+        d: data
+    };
+    let mut report = match rsgx_create_report(&ti, &report_data) {
+        Ok(report) => report,
+        Err(x) => {
+            println!("[-] rsgx_create_report() failed, error code: {:?}", x);
+            return None;
+        }
+    };
+
+    const RET_QUOTE_BUF_LEN : u32 = 9999;
+    let mut return_quote_buf : [u8; RET_QUOTE_BUF_LEN as usize] = [0;RET_QUOTE_BUF_LEN as usize];
+    let mut quote_len : u32 = 0;
+
+    let p_report = &mut report as *mut sgx_report_t;
+    let p_quote = return_quote_buf.as_mut_ptr();
+    let max_quote_len = RET_QUOTE_BUF_LEN;
+    let p_quote_len = &mut quote_len as *mut u32;
+
+    let result = unsafe {
+        ocall_generate_quote(
+            &mut rt as *mut sgx_status_t,
+            p_report,
+            p_quote, 
+            max_quote_len,
+            p_quote_len)
+    };
+    match result {
+        sgx_status_t::SGX_SUCCESS => (),
+        _ => return None,
+    };
+
+    Some(return_quote_buf[0..quote_len as usize].to_vec())
+
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn run_server() -> sgx_status_t {
```

Comment:
There is no "server" now, so the better name of this function maybe `init()`

Reply:
I am now planning to write the rest of the service process in this ecall, so I have planned a run_server function in advance.